weldx/util.py

```diff
@@ -1201,3 +1193,97 @@ def time_ref(self, value: pd.Timestamp):
                 self._obj.time.attrs["time_ref"] = value  # set new time_ref value
             else:
                 self._obj.time.attrs["time_ref"] = value
+
+
+class _Eq_compare_nested:
+    """Compares nested data structures like lists, sets, tuples, arrays, etc."""
+
+    compare_funcs = {
+        (np.ndarray, pint.Quantity, pd.Index): lambda x, y: np.all(x == y),
+        (xr.DataArray, xr.Dataset): lambda x, y: x.identical(y),
+    }
+
+    @staticmethod
+    def _compare(x, y):
+        # 1. treat special cases first
+        # 2. rely upon the fact, that comparison would fail,
+        #    if y is of an incompatible type.
+        for types, func in _Eq_compare_nested.compare_funcs.items():
+            if isinstance(x, types):
+                return func(x, y)
+            return x == y
+
+        return x == y
+
+    @staticmethod
+    def _enter(path, key, value):
```

Comment:
Maybe thats just me, but I like to have docstrings and return type hints on private class methods using the same style as for public methods. Makes it easier to understand for someone who might touch the code in the future (or reviews it 😛).

Reply:
These methods override functionality of boltons.iterutils, which doesn't use type annotations as well. I can add more documentation, if you feel like it is non yet sufficient (but I'd be delighted, if you could point me towards it, as I'm already too deeply involved with this code).

Reply:
I am not familiar with itertools. I can somehow anticipate what the inputs are, but what is the purpose of the second return value (bool)? I can google it, but the point is: 
I like having docstrings on internal functions in general until they are simple one-liners. We all have different knowledge bases and something that is super obvious to someone might be a riddle to another person (Including the future you 😛 ). I know that @CagtayFabry has a somewhat different opinion regarding this topic and we should probably discuss a general procedure. So no need to change something here and treat it as an annotation

Reply:
Thanks! I added more comments due to your remarks.
I totally agree to have some kind of documentation to aid the understanding of the code.
It is always a tradeoff how much to document. A quick line comment should sufficient for private methods in my opinion, if it is not too complex (e.g. weird behavior or lots of code). Large numpydoc styled strings for very simple functionality blows up the code and reading the actual code instead of the docstring would lead to a quicker understanding. This is of course not the case for something non-obvious or weird behavior, but then some comment lines should also do the trick.